### PR TITLE
Feature Request: Google Command (Fixed)

### PR DESCRIPTION
**Pull request by jpanahon**
### What are you trying to accomplish?
A way for students to google things whilst engaged in conversation on discord, this is very useful when somebody needs help with a specific programming issue.

### How are you accomplishing it?
By creating a command that displays google search results to Discord.
image

### Is there anything reviewers should know?
This is my second pull request for the same feature request, my original pull request had some changes I did not expect to have been made. 

Unlike my last pull request, I put the Google Custom Search API credentials in the command file itself instead of being environment variables.

To generate a key: https://developers.google.com/custom-search/v1/overview#api_key
To configure the search engine: https://programmablesearchengine.google.com/controlpanel/all

It is recommended you enable search the web for better and relevant results

If there is an issue you'd like me to address quickly just contact me on Discord. My discord is .morose.

- [x] Is it safe to rollback this change if anything goes wrong?
Yes, I only added a command, and changed config.example.yaml to enable the google command.

### CODE REVIEW

File: src/commands/google.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    const toWhom = interaction.options.getUser("to");
+    const embed = new EmbedBuilder()
+      .setColor(Colors.Blue)
+      .setAuthor({
+        name: `Requested by ${interaction.user.displayName}`,
+        iconURL: `${interaction.user?.avatarURL()}`,
+      })
+      .setTimestamp();
+
+    const search_key = "insert custom search api token";
```
Comment (kilbouri) [inline src/commands/google.ts]: Is there a reason these aren't provided by the `Config`?

Comment (jpanahon) [reply src/commands/google.ts]: > Is there a reason these aren't provided by the `Config`?

Didn't want to add more config variables, but if it is not an issue then I'll add them as environment variables like I did in my last pull request.

Comment (kilbouri) [reply src/commands/google.ts]: Its far more convenient than requiring a code change. Stuff like this should always be part of the config (or environment variables, but config is generally nicer).

File: src/commands/google.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            ephemeral: true,
+          });
+        }
+
+        return (await response.json()) as Results;
+      })
+      .then((data: Results) => {
+        let i: number = 0;
+        data.items.forEach((element: SearchResult) => {
+          links.push(`${++i}. [${element.title}](${element.link})`);
```
Comment (kilbouri) [inline src/commands/google.ts]: There's probably a special case we should cover here: no results. Maybe just a simple, `No results.` is sufficient.

File: src/commands/google.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    await fetch(url, {
+      method: "GET",
+      headers: {Accept: "*/*"},
+    })
+      .then(async (response: Response) => {
+        if (!response.ok) {
+          logger.info("No response.");
+          interaction.reply({
+            content: "Unable to return any queries.",
```
Comment (kilbouri) [inline src/commands/google.ts]: I would rephrase this a bit to make it clear that its an error on our side, not that there were no results.

File: src/commands/google.ts [hunk 4]
```diff
@@ -35,8 +36,8 @@ const googleModule: CommandType = {
       })
       .setTimestamp();
 
-    const search_key = "insert custom search api token";
-    const search_id = "insert custom search engine id";
+    const search_key = process.env.GOOGLE_SEARCH_KEY;
+    const search_id = process.env.GOOGLE_SEARCH_ID;
```
Comment (kilbouri) [inline src/commands/google.ts]: These can be made available on the `Config` object (see config.ts). This is the preferred way of acquiring configuration values (especially since those which are in the config YAML are not set as environment variables).

File: src/commands/google.ts [hunk 5]
```diff
@@ -68,9 +69,14 @@ const googleModule: CommandType = {
...
       .then((data: Results) => {
         let i: number = 0;
-        data.items.forEach((element: SearchResult) => {
-          links.push(`${++i}. [${element.title}](${element.link})`);
-        });
+
+        if (data.items.length == 0) {
+          links.push("No results.");
+        } else {
+          data.items.forEach((element: SearchResult) => {
```
Comment (kilbouri) [inline src/commands/google.ts]: If `items` is an array, this code is more expressive using `links = items.map((element) => ...);` instead.

File: src/commands/google.ts [hunk 6]
```diff
@@ -71,11 +71,12 @@ const googleModule: CommandType = {
         let i: number = 0;
 
         if (data.items.length == 0) {
-          links.push("No results.");
+          links[0] = "No results.";
```
Comment (kilbouri) [inline src/commands/google.ts]: Is this going to error out? The array doesn't contain any items so `[0]` is out of bounds.

Also, since you're using `map`, you can do something like
```js
let response;

if (no items) {
	response = "No results";
} else {
	response = items.map(...).join('\n')
}
```
or similar

**Review by kilbouri — CHANGES_REQUESTED** [review]

**Review by kilbouri — APPROVED** [review]
Assuming you've run and tested this current version, looks good to me!

File: src/commands/google.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        .setName("query")
+        .setDescription("What you want to Google.")
+        .setRequired(true)
+    )
+    .addUserOption((option) =>
+      option
+        .setName("to")
+        .setDescription("The user you want to direct this to.")
+    ),
+  execute: async (interaction: ChatInputCommandInteraction<CacheType>) => {
```
Comment (BorhanSaflo) [inline src/commands/google.ts:38]: Wrap the the code inside of the execute function with a try/catch block so that if anything was to go wrong while executing this command, it wont crash the bot.

File: src/commands/google.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+      const search_key = Config.google_search_key;
+      const search_id = Config.google_search_id;
+      const url = `https://www.googleapis.com/customsearch/v1?key=${search_key}&cx=${search_id}&q=${query}`;
+
+      let response: string = "";
+
+      const res = await fetch(url, {
+        method: "GET",
+        headers: {Accept: "*/*"},
```
Comment (kilbouri) [inline src/commands/google.ts]: Its just occurred to me that since you're expecting JSON (see `res.json()` below, lol) you should specify that you will only accept `application/json`.

File: src/commands/google.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+interface Results {
+  items: SearchResult[];
+}
+
+const googleModule: CommandType = {
+  data: new SlashCommandBuilder()
+    .setName("google")
+    .setDescription("Google something and return the top 10 results.")
+    .addStringOption((option) =>
```
Comment (BorhanSaflo) [inline src/commands/google.ts:26]: As of right now, there is no max for the query length, the command fails if the query is too long. You can set a max by adding ``.setMaxLength(n)``

**Review by BorhanSaflo — CHANGES_REQUESTED** [review]

**Review by BorhanSaflo — APPROVED** [review]
